lib/puppet/functions/libkv/delete.rb

```diff
@@ -1,86 +1,117 @@
+  #
+  #     * When set to a non-empty string, it is prepended to the key used in
+  #       the backend operation.
+  #     * Should only be set to an empty string when the key being accessed is
+  #       truly global.
+  #     * Defaults to the Puppet environment for the node.
+  #
+  #   * `resource`: String.  Name of the Puppet resource initiating this libkv
+  #     operation
+  #
+  #     * Required when `backend` is not specified and you want to be able
+  #       to use more than the `default` backend.
+  #     * String should be resource as it would appear in the catalog or
+  #       some application grouping id
+  #
+  #       * 'Class[<class>]' for a class, e.g.  'Class[Mymodule::Myclass]'
+  #       * '<Defined type>[<instance>]' for a defined type instance, e.g.,
+  #         'Mymodule::Mydefine[myinstance]'
+  #
+  #     * Catalog resource id cannot be reliably determined automatically.
+  #       Appropriate scope is not necessarily available when a libkv function
+  #       is called within any other function.  This is problematic for heavily
+  #       used Puppet built-in functions such as `each`.
+  #
+  #  * `softfail`: Boolean. Whether to ignore libkv operation failures.
   #
-  # @return [Any] The value in the underlying backing store
+  #    * When `true`, this function will return a result even when the operation
+  #      failed at the backend.
+  #    * When `false`, this function will fail when the backend operation failed.
+  #    * Defaults to `false`.
   #
+  # @raise [ArgumentError] If the key or merged backend config is invalid
+  #
+  # @raise [LoadError] If the libkv adapter cannot be loaded
+  #
+  # @raise [RuntimeError] If the backend operation fails, unless 'softfail' is
+  #   `true` in the merged backend options.
+  #
+  # @return [Boolean] `true` when backend operation succeeds; `false` when the
+  #   backend operation fails and 'softfail' is `true` in the merged backend
+  #   options
   #
   dispatch :delete do
-    param 'Hash', :parameters
+    required_param 'String[1]', :key
+    optional_param 'Hash',      :options
   end
 
+  def delete(key, options={})
+    # key validation difficult to do via a type alias, so validate via function
+    call_function('libkv::validate_key', key)
 
+    # add libkv 'extension' to the catalog instance as needed
+    call_function('libkv::add_libkv')
 
-  
-    dispatch :delete_v1 do
-    
-      
-        param "String", :parameters
-      
-    
-    end
-    def delete_v1(key)
-     params = {}
-     
-      
-        params['key'] = key
-      
-    
-    delete(params)
-    end
-  
-
-def delete(params)
-    nparams = params.dup
-    if (closure_scope.class.to_s == 'Puppet::Parser::Scope') 
+    # determine backend configuration using options, `libkv::options`,
+    # and the list of backends for which plugins have been loaded
+    begin
+      resource = options.has_key?('resource') ?  options['resource'] : 'unknown'
```

Comment:
I don't know why we would use the word `unknown` somewhere, but should we change this to something like `__libkv_unknown__` just to reasonably eliminate any chance of overlap?

It would also potentially help with debugging.

Reply:
It will never show up in debugging.  The code will look for 'default' if there is no match.  If 'default' is not found, the code will fail.

Reply:
@lnemsick-simp I meant more like `grep`-style debugging